js/src/functions/upload.ts

```diff
@@ -374,18 +445,44 @@ function formatNameAndSlug(pieces: string[]) {
   };
 }
 
-async function loadProjectId(projectName: string): Promise<string> {
-  const response = await _internalGetGlobalState()
-    .appConn()
-    .post_json("api/project/register", {
-      project_name: projectName,
-    });
+export class ProjectNameIdMap {
+  private nameToId: Record<string, string> = {};
+  private idToName: Record<string, string> = {};
+
+  async getId(projectName: string): Promise<string> {
+    if (!(projectName in this.nameToId)) {
+      const response = await _internalGetGlobalState()
+        .appConn()
+        .post_json("api/project/register", {
+          project_name: projectName,
+        });
+
+      const result = z
+        .object({
+          project: projectSchema,
+        })
+        .parse(response);
 
-  const result = z
-    .object({
-      project: projectSchema,
-    })
-    .parse(response);
+      const projectId = result.project.id;
 
-  return result.project.id;
+      this.nameToId[projectName] = projectId;
+      this.idToName[projectId] = projectName;
+    }
+    return this.nameToId[projectName];
+  }
+
+  async getName(projectId: string): Promise<string> {
+    if (!(projectId in this.idToName)) {
+      const response = await _internalGetGlobalState()
+        .appConn()
+        .post_json("api/project/get", {
+          id: projectId,
+        });
+      const result = z.array(projectSchema).parse(response);
```

Comment:
You can use the `.nonempty()` constraint to make zod check that the array is nonempty too.

Reply:
done